aksparse/client/aksparse/sparse.py

```diff
@@ -442,48 +474,54 @@ def dot(self, input):
-            arr_ids = resp.split("+")
-            data = ak.create_pdarray(f"created {arr_ids[2]}")
-            indices = ak.create_pdarray(f"created {arr_ids[1]}")
-            ind_ptr = ak.create_pdarray(f"created {arr_ids[0]}")
-            return Csr(data, indices, ind_ptr, self.shape, finished_vals=True)
-        else:
+    def spgemm(self, other):
+        """ Sparse General Matrix Multiplication """
+        if os.environ.get('ARKOUDA_INIT') == "1":
             try:
                 assert(self.shape[1] == other.shape[0])
             except:
                 error_msg = f"array size mismatch"
                 raise AttributeError(error_msg)
+            
+            # identify number of multiplications needed
             starts = other.indptr[self._gb_col_row.unique_keys[0]]
             ends = other.indptr[self._gb_col_row.unique_keys[0] + 1]
-
             fullsize = (ends - starts).sum()
             nonzero = (ends > starts)
-
             lengths = (ends - starts)
             segs = ak.cumsum(lengths) - lengths
+
+            fsegs = segs[nonzero]
+            fstarts = starts[nonzero]
+            fends = ends[nonzero]
+
             totlen = lengths.sum()
             slices = ak.ones(totlen, dtype=ak.akint64)
-            diffs = ak.concatenate((ak.array([starts[0]]), starts[1:] - starts[:-1] - (lengths[:-1] - 1)))
-            slices[segs] = diffs
-            fullsegs, ranges =  segs, ak.cumsum(slices)
+            diffs = ak.concatenate((ak.array([fstarts[0]]), fstarts[1:] - fends[:-1] + 1))
 
-            # fullsegs, ranges = ak.gen_ranges(starts[nonzero], ends[nonzero])
+            # Set up arrays for multiplication
+            slices[fsegs] = diffs
+            fullsegs, ranges = fsegs, ak.cumsum(slices)
             fullBdom = other._gb_row_col.unique_keys[1][ranges]
             fullAdom = ak.broadcast(fullsegs, self._gb_col_row.unique_keys[1][nonzero], fullsize)
-
             fullBval = other.data[ranges]
             fullAval = ak.broadcast(fullsegs, self.data[nonzero], fullsize)
-
             fullprod = fullAval * fullBval
+
+            # GroupBy indices and perform aggregate sum
             proddomGB = ak.GroupBy([fullAdom, fullBdom])
             result = proddomGB.sum(fullprod)
-            
-            return Csr(result[1], result[0][1], result[0][0], shape = (self.shape[0], other.shape[1]), no_chpl=True)
\ No newline at end of file
+            return Csr(result[1], result[0][0], result[0][1], shape = (self.shape[0], other.shape[1]))
+        else:
+            resp = cast(str, ak.client.generic_msg(cmd="spgemm", args={"self_data": self.data, "other_indptr" : other.indptr, "other_data" : other.data, "self_shape_width": self.shape[0], "self_shape_height" : self.shape[1], "other_shape_width" : other.shape[0], "other_shape_height" : other.shape[1], "self_gb_rep" : self._gb_col_row, "other_gb_rep" : other._gb_row_col, "cname": self._cname, "rname": self._rname}));
+            arr_ids = resp.split("+")
```

Comment:
We should be able to support returns from the server also being JSON format. We are trying to move all messages to JSON formatting. Let me know if you need help with this.